### ISSUE TITLE
implicit parameter conversion

### ISSUE THREAD

**Issue by aeantipov** [ALPSCore/ALPSCore#166]
this code now fails (g++-4.9.2):
```
alps::params p(argc, (const char**)argv);
p.define<std::string>("output",        "output.h5", "output file");
alps::hdf5::archive ar(p["output"], "w");
```
with the error:
```
Option 'output': Attempt to assign an option_type object containing a value of type "std::string" to a value of an incompatible type "N5boost10filesystem4pathE"
```
The reason is that the alps::hdf5::archive requires boost::filesystem::path, which can be converted from std::string. But two types are different and the parameter can not be extracted. 
What shall we do with this and similar errors (like bool->int, float->double, etc)?

**Comment by galexv** [comment]
I agree with an implicit conversion to a type to which there is a conversion defined.
But are we fine with conversions like `int x=par["double_param"];` ? Currently it throws an exception, because `int` is narrower than `double`. 
I would suggest to prohibit narrowing conversions among primitive types, and allow other conversions, including user-defined ones.

**Comment by egull** [comment]
Will be dealt with in #195
